src/error/metadata/code.rs

```diff
@@ -1,11 +1,147 @@
 use std::fmt::{self, Display};
-
+use crate::anyhow;
+use std::collections::BTreeMap;
+use std::iter::FromIterator;
 /// `Code` contains the error codes associated with specific errors.
-#[derive(Debug)]
-pub enum Code {}
+#[derive(Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub enum Code {
+    E001,
+    E002,
+    E003,
+    E004,
+    E005,
+    E006,
+    E007,
+    E008,
+    E009,
+    E010,
+    E011,
+    E012,
+    E013,
+    E014,
+    E015,
+    E016,
+    E017,
+    E018,
+    E019,
+    E020,
+    E021,
+    E022,
+    E023,
+    E024,
+    E025,
+    E026,
+    E027,
+    E028,
+}
 
 impl Display for Code {
     fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
         write!(formatter, "{:?}", &self)
     }
 }
+
+/// for converting from a text representation of the code to the code itself,
+/// useful for `explain` command input to display error explanations.
+impl std::convert::TryFrom<&str> for Code {
+    type Error = anyhow::Error;
+    fn try_from(code: &str) -> Result<Self, Self::Error> {
+        match code {
+            "E001" => Ok(Code::E001),
+            "E002" => Ok(Code::E002),
+            "E003" => Ok(Code::E003),
+            "E004" => Ok(Code::E004),
+            "E005" => Ok(Code::E005),
+            "E006" => Ok(Code::E006),
+            "E007" => Ok(Code::E007),
+            "E008" => Ok(Code::E008),
+            "E009" => Ok(Code::E009),
+            "E010" => Ok(Code::E010),
+            "E011" => Ok(Code::E011),
+            "E012" => Ok(Code::E012),
+            "E013" => Ok(Code::E013),
+            "E014" => Ok(Code::E014),
+            "E015" => Ok(Code::E015),
+            "E016" => Ok(Code::E016),
+            "E017" => Ok(Code::E017),
+            "E018" => Ok(Code::E018),
+            "E019" => Ok(Code::E019),
+            "E020" => Ok(Code::E020),
+            "E021" => Ok(Code::E021),
+            "E022" => Ok(Code::E022),
+            "E023" => Ok(Code::E023),
+            "E024" => Ok(Code::E024),
+            "E025" => Ok(Code::E025),
+            "E026" => Ok(Code::E026),
+            "E027" => Ok(Code::E027),
+            "E028" => Ok(Code::E028),
+            _ => Err(anyhow!("Invalid error code. Error codes are in the format `E###`"))
+        }
+    }
+}
+
+impl Code {
+    // builds a BTreeMap of every possible code and its explanation, so we can
+    // access from the `explain` function and get a single one OR so we can
+    // iterate over them in `explain_all` for creating a docs page
+    fn explanations() -> BTreeMap<Code, String>{
+        let contents = vec![
```

Comment:
and for this we can likely use strum's EnumIterator https://docs.rs/strum/0.20.0/strum/derive.EnumIter.html

Reply:
I'm having  a little trouble with this, I'm not sure it'll work. Since the include_str macro is run at buildtime, it needs a string literal, so iterating over the enum and trying to dynamically load the paths doesn't seem to be possible, unless I'm missing something. Any ideas?

Reply:
hmmmmmm that's a good point! probably fine to leave it as is, though i'll cc @lrlna just in case they have any suggestions there